prover/src/matrix/row_matrix.rs

```diff
@@ -0,0 +1,200 @@
+    }
+
+    /// Returns a reference to a row at the specified index in this matrix.
+    ///
+    /// # Panics
+    /// Panics if the specified row index is out of bounds.
+    pub fn get_row(&self, row_idx: usize) -> &[E] {
+        assert!(row_idx < self.num_rows());
+        let start = row_idx * self.row_len;
+        &self.data[start..start + self.row_len]
+    }
+
+    /// Returns the data in this matrix as a slice of field elements.
+    pub fn as_data(&self) -> &[E] {
+        &self.data
+    }
+}
+
+// HELPER FUNCTIONS
+// ================================================================================================
+
+/// Returns a vector of offsets for the specified number of rows. The offsets are computed as
+/// `domain_offset^i` for `i` in `[0, num_rows)`. The first offset is always 1.
+fn get_offsets<E>(num_rows: usize, domain_offset: E::BaseField) -> Vec<E::BaseField>
+where
+    E: FieldElement,
+{
+    // create a vector to hold the offsets.
+    let mut offsets = Vec::with_capacity(num_rows);
+
+    // the first offset is always 1.
+    offsets.push(E::BaseField::ONE);
+
+    // compute the remaining offsets.
+    for i in 1..num_rows {
+        offsets.push(offsets[i - 1] * domain_offset);
+    }
+
+    offsets
+}
+
+/// Creates a vector of uninitialised segments to hold the result. The number of segments is
+/// equal to the number of columns in the matrix divided by the ARR_SIZE. Each segment is
+/// initialised with a vector of uninitialised arrays of length `num_rows * blowup_factor`.
+///
+/// # Panics
+/// Panics if the number of columns in the matrix is not a multiple of ARR_SIZE.
+fn allocate_segments<E>(num_rows: usize, row_width: usize, blowup_factor: usize) -> Vec<Segment<E>>
+where
+    E: FieldElement,
+{
+    assert!(
+        row_width % ARR_SIZE == 0,
+        "number of columns must be a multiple of ARR_SIZE"
+    );
+    let outer_vec_len = row_width / ARR_SIZE;
+
+    // create a vector of uninitialised segments to hold the result.
+    // SAFETY: we are creating a vector of uninitialised segments, and each segment is
+    // initialised with a vector of uninitialised arrays of length `num_rows * blowup_factor`.
+    // This is safe because we are not reading from the vectors, and we will initialise
+    // the vectors before reading from them.
+    let outer_vec: Vec<Segment<E>> = (0..outer_vec_len)
+        .map(|_| Segment::new(unsafe { uninit_vector::<[E; ARR_SIZE]>(num_rows * blowup_factor) }))
+        .collect();
+
+    outer_vec
+}
+
+/// Prepares a segment for evaluation by multiplying each element in the segment by the
+/// corresponding offset.
+fn prepare_segment<E>(
+    segment: &mut Segment<E>,
+    polys: &Matrix<E>,
+    seg_idx: usize,
+    offsets: &[E::BaseField],
+) where
+    E: FieldElement,
+{
+    (seg_idx * ARR_SIZE..(seg_idx + 1) * ARR_SIZE).for_each(|col_idx| {
+        // get the column from the polys matrix.
+        let row = polys.get_column(col_idx);
```

Comment:
Is the variable naming here correct? Should it be `column` rather than `row`?

Reply:
yes, thanks